_posts/2018-4-11-IO-Polling.md

```diff
@@ -0,0 +1,96 @@
+thread, it is almost always better to use kernel-space polling. The user-space
+polling implementation is left in place for platforms that do not support
+parallel update of the kernel-space pollset. Also user-space polling is used
+for individual file descriptors when they cannot be put in a kernelspace pollset
+for some reason.
+
+## Poll-threads and Poll-sets
+
+OTP 21 introduces two new configuration parameters: +IOt and +IOp.
+
+### Configure +IOt
+
++IOt controls the number of threads that are used to deliver events. The default
+is 1 and it should be enough for most applications. However on very busy
+systems with many cuncurrent connection it could be beneficial to increase this.
+One way to get an indication of whether your system could benefit from it is
+by using [msacc](http://erlang.org/doc/man/msacc.html). If you turn it on briefly
+and when examining the `msacc:print()` output notice that sleep time
+of the the thread type `poll` is low, the system may benefit from increasing the
+number of polling threads.
+
+```
+Eshell V9.3  (abort with ^G)
+1> msacc:start(10000),msacc:print().
+Average thread real-time    : 10000410 us
+Accumulated system run-time :      937 us
+Average scheduler run-time  :      897 us
+
+        Thread      aux check_io emulator       gc    other     port    sleep
+
+Stats per thread:
+     async( 0)    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+       aux( 1)    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+dirty_cpu_( 1)    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+dirty_io_s( 1)    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+      poll( 0)    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+ scheduler( 1)    0.00%    0.00%    0.00%    0.00%    0.01%    0.00%   99.99%
+
+Stats per type:
+         async    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+           aux    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+dirty_cpu_sche    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+dirty_io_sched    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+          poll    0.00%    0.00%    0.00%    0.00%    0.00%    0.00%  100.00%
+     scheduler    0.00%    0.00%    0.00%    0.00%    0.01%    0.00%   99.99%
+```
+
+In the example above the poll thread is sleeping for 100% of the time so no need to
+increase the number of poll theads.
+
+### Configure +IOp
+
++IOp controls the number of pollsets used to put the file descriptors in. This
+options defaults to 1, and it should be very very rare for any system to benefit
+from this. The only time so far that I have seen it to be beneficial is when the
+kernel-space poll implementation does not scale well when accessed in parallel
+by multiple threads. So if you run [perf top](http://man7.org/linux/man-pages/man1/perf-top.1.html)
+(or something sililar) on your system and notice that a lot of time is spent
```

Comment:
In the last paragraph it says . 
This options defaults to 1, and it should be very very rare for any system to benefit from this. 

Change to... Should be very rare for any system to benefit from changing this.